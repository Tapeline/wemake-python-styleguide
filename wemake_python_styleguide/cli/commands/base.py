"""Contains files common for all wps commands."""

from abc import ABC, abstractmethod


class AbstractCommand(ABC):
    """ABC for all commands."""

    @abstractmethod
    def run(self, args) -> int:
        """Run the command."""
        raise NotImplementedError
