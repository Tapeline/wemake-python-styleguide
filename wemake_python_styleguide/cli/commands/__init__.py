"""Contains all files related to wps console commands."""
