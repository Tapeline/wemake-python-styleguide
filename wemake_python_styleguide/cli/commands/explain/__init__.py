"""Contains files related to wps explain command."""
