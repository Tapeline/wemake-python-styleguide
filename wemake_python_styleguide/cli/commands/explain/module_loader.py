import importlib
from collections.abc import Collection
from contextlib import suppress
from pathlib import Path
from types import ModuleType
from typing import Final

_VIOLATION_MODULE_BASE: Final = 'wemake_python_styleguide.violations'


def get_violation_submodules() -> Collection[ModuleType]:
    """Get all possible violation submodules."""
    return _safely_get_all_submodules(_VIOLATION_MODULE_BASE)


def _safely_get_all_submodules(module_name: str) -> Collection[ModuleType]:
    """Get all submodules of given module. Ignore missing."""
    submodule_names = _get_all_possible_submodule_names(module_name)
    modules = []
    for submodule_name in submodule_names:
        # just in case if there are some bad module paths
        # (which generally should not happen)
        with suppress(ModuleNotFoundError):
            modules.append(importlib.import_module(submodule_name))
    return modules


def _get_all_possible_submodule_names(module_name: str) -> Collection[str]:
    """Get .py submodule names listed in given module."""
    root_module = importlib.import_module(module_name)
    root_paths = root_module.__path__
    names = []
    for root in root_paths:
        names.extend([
            f'{module_name}.{name}'
            for name in _get_all_possible_names_in_root(root)
        ])
    return names


def _get_all_possible_names_in_root(root: str) -> Collection[str]:
    """Get .py submodule names listed in given root path."""
    return [
        path.name.removesuffix('.py')
        for path in Path(root).glob('*.py')
        if '__' not in path.name  # filter dunder files like __init__.py
    ]
