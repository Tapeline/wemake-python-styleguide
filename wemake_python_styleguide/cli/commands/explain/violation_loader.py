"""Provides tools to extract violation info."""

import importlib
import inspect
from collections.abc import Collection, Mapping
from types import ModuleType
from typing import Final, final

from attrs import frozen

from wemake_python_styleguide.violations.base import BaseViolation

_VIOLATION_SUBMODULES: Final = (
    'best_practices',
    'complexity',
    'consistency',
    'naming',
    'oop',
    'refactoring',
    'system',
)
_VIOLATION_MODULE_BASE: Final = 'wemake_python_styleguide.violations'


@final
@frozen
class ViolationInfo:
    """Contains violation info."""
    identifier: str
    fully_qualified_id: str
    code: int
    docstring: str
    section: str


def _is_a_violation(class_object) -> bool:
    """Check if class is a violation class."""
    return (
        issubclass(class_object, BaseViolation) and
        hasattr(class_object, 'code')  # Not all subclasses have code
    )


def _get_violations_of_submodule(
    module: ModuleType,
) -> Collection[type[BaseViolation]]:
    """Get all violation classes of defined module."""
    return [
        class_
        for name, class_ in inspect.getmembers(module, inspect.isclass)
        if _is_a_violation(class_)
    ]


def _create_violation_info(
    class_object, submodule_name: str, submodule_path: str
) -> ViolationInfo:
    """Create violation info DTO from violation class and metadata."""
    return ViolationInfo(
        identifier=class_object.__name__,
        fully_qualified_id=f'{submodule_path}.{class_object.__name__}',
        code=class_object.code,
        docstring=class_object.__doc__,
        section=submodule_name,
    )


def _get_all_violations() -> Mapping[int, ViolationInfo]:
    """Get all violations inside all defined WPS violation modules."""
    all_violations = {}
    for submodule_name in _VIOLATION_SUBMODULES:
        submodule_path = f'{_VIOLATION_MODULE_BASE}.{submodule_name}'
        violations = _get_violations_of_submodule(
            importlib.import_module(submodule_path)
        )
        for violation in violations:
            all_violations[violation.code] = _create_violation_info(
                violation,
                submodule_name,
                submodule_path,
            )
    return all_violations


def get_violation(code: int) -> ViolationInfo | None:
    """Get a violation by its integer code."""
    violations = _get_all_violations()
    if code not in violations:
        return None
    return violations[code]
