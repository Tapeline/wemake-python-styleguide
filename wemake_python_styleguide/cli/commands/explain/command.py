"""Contains command implementation."""

from wemake_python_styleguide.cli.commands.base import AbstractCommand
from wemake_python_styleguide.cli.commands.explain import (
    message_formatter,
    violation_loader,
)


def _clean_violation_code(violation_str: str) -> int:
    """
    Get int violation code from str violation code.

    Args:
        violation_str: violation code expressed as string
                       WPS412, 412 - both acceptable

    Returns:
        integer violation code

    Throws:
        ValueError: violation str is not an integer (except WPS prefix).
    """
    violation_str = violation_str.removeprefix('WPS')
    try:
        return int(violation_str)
    except ValueError:
        return -1


class ExplainCommand(AbstractCommand):
    """Explain command impl."""

    def run(self, args) -> int:
        """Run command."""
        code = _clean_violation_code(args.violation_code)
        violation = violation_loader.get_violation(code)
        if violation is None:
            self.writer.write_err('Violation not found')
            self.writer.flush()
            return 1
        message = message_formatter.format_violation(violation)
        self.writer.write_out(message)
        self.writer.flush()
        return 0
