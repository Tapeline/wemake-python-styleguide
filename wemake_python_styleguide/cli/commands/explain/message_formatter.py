"""Provides tools for formatting explanations."""

from typing import Final

from wemake_python_styleguide.cli.commands.explain.violation_loader import (
    ViolationInfo,
)

_DOCS_URL: Final = 'https://pyflak.es/{0}'


def _clean_text(text: str) -> str:
    """
    Cleans provided text.

    Args:
        text: target text

    Returns:
        text with normalized newlines (CRs and CRLFs transformed to LFs).
    """
    return text.replace('\r\n', '\n').replace('\r', '\n')


def _replace_tabs(text: str, tab_size: int = 4) -> str:
    """Replace all tabs with defined amount of spaces."""
    return text.replace('\t', ' ' * tab_size)


def _get_whitespace_prefix(line: str) -> int | float:
    """Get length of whitespace prefix of string."""
    for char_index, char in enumerate(line):
        if char != ' ':
            return char_index
    return float('+inf')


def _get_greatest_common_indent(text: str) -> int:
    """Get the greatest common whitespace prefix length of all lines."""
    lines = text.split('\n')
    greatest_common_indent = float('+inf')
    for line in lines:
        greatest_common_indent = min(
            greatest_common_indent, _get_whitespace_prefix(line)
        )
    if isinstance(greatest_common_indent, float):
        greatest_common_indent = 0
    return greatest_common_indent


def _remove_indentation(text: str, tab_size: int = 4) -> str:
    """Remove excessive indentation."""
    text = _replace_tabs(_clean_text(text), tab_size)
    max_indent = _get_greatest_common_indent(text)
    return '\n'.join(line[max_indent:] for line in text.split('\n'))


def format_violation(violation: ViolationInfo) -> str:
    """Format violation information."""
    cleaned_docstring = _remove_indentation(violation.docstring)
    violation_url = _DOCS_URL.format(f'WPS{violation.code}')
    return (
        f'WPS{violation.code} ({violation.identifier})\n'
        f'{cleaned_docstring}\n'
        f'See at website: {violation_url}'
    )
