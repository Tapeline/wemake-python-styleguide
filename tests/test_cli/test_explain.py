"""Test that wps explain command works fine."""

import pytest

from wemake_python_styleguide.cli.commands.explain import (
    message_formatter,
    violation_loader,
)
from wemake_python_styleguide.violations.best_practices import (
    InitModuleHasLogicViolation,
)
from wemake_python_styleguide.violations.naming import (
    UpperCaseAttributeViolation,
)
from wemake_python_styleguide.violations.oop import BuiltinSubclassViolation


@pytest.mark.parametrize(
    'violation_params',
    [
        (115, UpperCaseAttributeViolation),
        (412, InitModuleHasLogicViolation),
        (600, BuiltinSubclassViolation),
    ],
)
def test_violation_getter(violation_params):
    """Test that violation loader can get violation by their codes."""
    violation_code, expected_class = violation_params
    violation = violation_loader.get_violation(violation_code)
    assert violation.code is not None
    assert violation.docstring == expected_class.__doc__


@pytest.mark.parametrize(
    'test_params',
    [
        ('  text\n  text\n  text', 'text\ntext\ntext'),
        ('  text\n\ttext\r\n  text', 'text\n  text\ntext'),
    ],
)
def test_indentation_removal(test_params):
    """Test that indentation remover works in different conditions."""
    input_text, expected = test_params
    actual = message_formatter._remove_indentation(input_text)  # noqa: SLF001
    assert actual == expected


violation_mock = violation_loader.ViolationInfo(
    identifier='Mock',
    code=100,
    docstring='docstring',
    fully_qualified_id='mock.Mock',
    section='mock',
)
violation_string = (
    'WPS100 (Mock)\n'
    'docstring\n'
    'See at website: https://wemake-python-styleguide.readthedocs.io/en/'
    'latest/pages/usage/violations/mock.html#mock.Mock'
)


def test_formatter():
    """Test that formatter formats violations as expected."""
    formatted = message_formatter.format_violation(violation_mock)
    assert formatted == violation_string
