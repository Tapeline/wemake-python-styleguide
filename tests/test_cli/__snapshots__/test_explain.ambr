# serializer version: 1
# name: test_command
  '''
  WPS123 â€” Forbid unused variables with multiple underscores.
  
  Reasoning:
      We only use ``_`` as a special definition for an unused variable.
      Other variables are hard to read. It is unclear why would one use it.
  
  Solution:
      Rename unused variables to ``_``
      or give it some more context with an explicit name: ``_context``.
  
  Example::
  
      # Correct:
      some_element, _next_element, _ = some_tuple()
      some_element, _, _ = some_tuple()
      some_element, _ = some_tuple()
  
      # Wrong:
      some_element, _, __  = some_tuple()
  
  .. versionadded:: 0.12.0
  
  See at website: https://pyflak.es/WPS123
  
  '''
# ---
# name: test_command_on_not_found[wps explain 10000]
  '''
  Violation not found
  
  '''
# ---
# name: test_command_on_not_found[wps explain NOT_A_CODE]
  '''
  Violation not found
  
  '''
# ---
# name: test_command_on_not_found[wps explain WPS10000]
  '''
  Violation not found
  
  '''
# ---
# name: test_no_command_specified
  '''
  Command not specified. Usage: wps help
  
  '''
# ---
